pkg/filter/filter.go

```diff
@@ -72,22 +78,93 @@ func (h authHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	logrus.Debugf("Impersonating user %v, groups %v", user, groups)
+	var impersonateUser bool
+	var impersonateGroup bool
 
-	req.Header.Set("Impersonate-User", user)
+	reqUser := req.Header.Get("Impersonate-User")
+	var reqGroup []string
+	if g, ok := req.Header["Impersonate-Group"]; ok {
+		reqGroup = g
+	}
+
+	// If there is an impersonate header, the incoming request is attempting to
+	// impersonate a different user, verify the token user is authz to impersonate
+	if reqUser != "" && reqUser != user {
+		canDo, err := h.checkUserCanImpersonate(user, "user")
+		if err != nil {
+			util.ReturnHTTPError(rw, req, 401, err.Error())
+			return
+		} else if !canDo {
+			util.ReturnHTTPError(rw, req, 401, "not allowed to impersonate")
+			return
+		}
+		impersonateUser = true
+	}
+
+	if len(reqGroup) > 0 && !groupsEqual(reqGroup, groups) {
+		canDo, err := h.checkUserCanImpersonate(user, "group")
+		if err != nil {
+			util.ReturnHTTPError(rw, req, 401, err.Error())
+			return
+		} else if !canDo {
+			util.ReturnHTTPError(rw, req, 401, "not allowed to impersonate")
+			return
+		}
+		impersonateGroup = true
+	}
 
-	req.Header.Del("Impersonate-Group")
-	for _, group := range groups {
-		req.Header.Add("Impersonate-Group", group)
+	// Not impersonating either, set to the token users creds
+	if !impersonateUser && !impersonateGroup {
+		req.Header.Set("Impersonate-User", user)
+		req.Header.Del("Impersonate-Group")
+		for _, group := range groups {
+			req.Header.Add("Impersonate-Group", group)
+		}
+	} else if impersonateUser && !impersonateGroup { // Impersonating user, drop all groups
+		req.Header.Del("Impersonate-Group")
+	} else if !impersonateUser && impersonateGroup { // Impersonating groups, drop the user
+		req.Header.Del("Impersonate-User")
 	}
 
+	logrus.Debugf("Requesting user: %v, Requesting groups: %v Impersonate user: %v Impersonate group: %v",
+		user, groups, impersonateUser, reqGroup)
+
 	aw := rw
 	if h.auditWriter != nil {
 		auditID := k8stypes.UID(uuid.NewRandom().String())
-		h.auditWriter.LogRequest(req, auditID, authed, req.Header.Get("Content-Type"), user, groups)
+		h.auditWriter.LogRequest(req, auditID, authed, req.Header.Get("Content-Type"), user, reqUser, groups, reqGroup)
 
 		aw = &wrapWriter{auditID, rw, h.auditWriter, http.StatusOK}
 	}
 
 	h.next.ServeHTTP(aw, req)
 }
+
+func (h authHandler) checkUserCanImpersonate(user, resource string) (bool, error) {
+	review := authV1.SubjectAccessReview{
+		Spec: authV1.SubjectAccessReviewSpec{
+			User: user,
+			ResourceAttributes: &authV1.ResourceAttributes{
+				Verb:     "impersonate",
+				Resource: resource,
+			},
+		},
+	}
+
+	result, err := h.k8sClient.AuthorizationV1().SubjectAccessReviews().Create(&review)
```

Comment:
is SubjectAccessReviews a real kubernetes resource and should we delete it after it has been used?

Reply:
It is not real, nothing to delete.

Reply:
`h.k8sClient` points at the management k8s. Doing the subjectAccessReview against that is not correct/appropriate because a user typically wont be have impersonation control in the mgmt plane, but will have it in any cluster that they have the owner role in.

The following scenario(s) must work:
- I am owner of Cluster-A, a plain member of Cluster-B, and NOT an admin of the rancher mgmt plane cluster.
- I create a kubeconfig for Cluster-A and can do --as or --as-group against it because I am an admin 
- I create a kubeconfig for Cluster-B and cannot --as or --as-group against it because I am not an admin

Reply:
Possible solution:
`vendor/github.com/rancher/norman/store/proxy/proxy_store.go:127` is what copies the headers from the original request that the user made to the proxy request that actually goes to the user's k8s api. You could enhance this area of the code to look for new headers called something like "X-Impersonate-User"* and "X-Impersonate-Group"* and if those headers are present, make the subjectAccessReview request to the user k8s api and if successful, use set the x-impersonate-* headers to populate the proxy request instead of the normal ones.